Tasks/VsTest/VSTest.ts

```diff
@@ -0,0 +1,501 @@
+    
+    //Write-Host "##vso[task.logissue type=warning;TaskName=VSTest]"
+
+    var sourcesDirectory = tl.getVariable('System.DefaultWorkingDirectory');
+    var testAssemblyFiles = [];
+    if(testAssembly.indexOf('*') >= 0 || testAssembly.indexOf('?') >= 0) {
+        tl.debug('Pattern found in solution parameter.');
+        var excludeTestAssemblies = [];
+        var allFiles = tl.find(sourcesDirectory);
+        var testAssemblyFilters = testAssembly.split(';');
+        testAssemblyFilters.forEach(function(testAssemblyFilter){
+            if(testAssemblyFilter.startsWith("-:")){
+                if(testAssemblyFilter.indexOf('*') >= 0 || testAssemblyFilter.indexOf('?') >= 0){                    
+                    excludeTestAssemblies = excludeTestAssemblies.concat(getFilteredFiles(testAssemblyFilter.substr(2), allFiles));
+                }
+                else{
+                    excludeTestAssemblies.push(testAssemblyFilter.substr(2));
+                }
+            }
+            else if(testAssemblyFilter.indexOf('*') >= 0 || testAssemblyFilter.indexOf('?') >= 0){
+                testAssemblyFiles = testAssemblyFiles.concat(getFilteredFiles(testAssemblyFilter, allFiles));
+            }
+            else{
+                testAssemblyFiles.push(testAssemblyFilter);
+            }
+        });
+        testAssemblyFiles = removeArrayFromArray(testAssemblyFiles, excludeTestAssemblies);              
+    }
+    else{
+        tl.debug('No Pattern found in solution parameter.');
+        testAssembly.replace(';;', "`0")  // Borrowed from Legacy File Handler
+        var assemblies = testAssembly.split(';');
+        assemblies.forEach(function(assembly) {
+            testAssemblyFiles.push(assembly);
+        });
+    }
+
+    if (testAssemblyFiles && testAssemblyFiles.length != 0) { 
+        testAssemblyFiles = removeDuplicatesFromArray(testAssemblyFiles);  
+        var workingDirectory = path.join(sourcesDirectory, "..");     
+        getTestResultsDirectory(runSettingsFile, path.join(workingDirectory, 'TestResults')).then(function(resultsDirectory){
+            invokeVSTest(resultsDirectory).then(function(code){
+                try{
+                    publishTestResults(resultsDirectory);
+                    tl.exit(code);
+                }
+                catch(error){
+                    //Write-Host "##vso[task.logissue type=error;code=" $_.Exception.Message ";TaskName=VSTest]"
+                    throw error;
+                }
+            });
+        });        
+    }
+    else {
+        //Write-Host "##vso[task.logissue type=warning;code=002004;]"
+        tl.warning("No test assemblies found matching the pattern: " + testAssembly);
+    }
+}
+catch(error){
+    //Write-Host "##vso[task.logissue type=error;code=" $_.Exception.Message ";TaskName=VSTest]"
+    throw error;
+}
+
+function invokeVSTest(testResultsDirectory: string): Q.Promise<number> {
```

Comment:
I would suggest. normal functions should throw normal error/exceptions.
Main function can consolidate/log error based on the error message. One place/function for all errors